### PR TITLE
Online HCAL DQM adding a new application

### PR DESCRIPTION
**Pull request by vkhristenko**
This is a test PR based on https://github.com/cms-sw/cmssw/pull/13565
to test the addition of a new application to the Online HCAL DQM Workflow

VK

### CODE REVIEW

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
please submit to 80x as well

**Comment by deguio** [comment]
please test

**Comment by vanbesien** [comment]
please test

**Comment by vkhristenko** [comment]
@deguio decreased memory for offline
now, the modules running in offline should be in total <~6-7MB

VK

**Comment by vanbesien** [comment]
please test

**Comment by deguio** [comment]
ciao viktor,
the code looks ok to me. the only question I have is if all those utility classes you developed (electronics mapping etc) under the DQM hat, could be of general help in the HCAL reco code.
F.

**Comment by vkhristenko** [comment]
Yo @deguio 
So, I don't know what to say - can we just skip this question by any chance?
There are some analogies, that are already implemented, however HCAL Core software is
1) not good performance-wise (use of HCAL electornics map decreases significantly look up times - affects the statistics in Online).
2) not flexible - for instance this famous switch from VME to uTCA. If you have both Electronics sitting in the electronics HCAL map, it won't work properly...

I prefer not to discuss details, but like David already said, after putting templates and some code standartization, a lot of things from here can be pushed up.

VK

**Comment by deguio** [comment]
+1
let's merge it then. I would try to 'advertise' the work you did so that other HCAL code developers who might be interested are informed. the perfect forum would be the hcal readiness review currently ongoing.

F.

**Comment by vkhristenko** [comment]
sounds good. We will touch upon that in a bit today I guess...

Vk

**Comment by vkhristenko** [comment]
I'm suspending this PR for the moment!
I have found an issue in my logic for Offline processing and have been recoding it since Friday. I need a few more days - prolly just 1 to finish tests.. I know I'm running late - doing my best!

VK

**Comment by vkhristenko** [comment]
I'm reopening this PR as HCAL DQM CMSSW part is ready.

After the Readiness Reaview a few problems were observed + there were requests which I wante to finish off. Changes:

1) HcalTasks also contains "src" to generate a lib from that to be used by plugins. 
2) Summary generation - or Status generation was completely rewritten.

DQM OFfline Sequences were updated. everything is tested with whiterabbit... and works good.
HCAL DQM is ready now! I know I've said that before, but I think we are ready now.

VK

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
+1

**Comment by deguio** [comment]
@davidlange6 
could we have this one merged please?
thanks,
F.

@dmitrijus @vanbesien

**Comment by abdoulline** [comment]
I double Federico's request to David   @davidlange6 
Then we could hopefully proceed with (back-ported to 80X) 

 https://github.com/cms-sw/cmssw/pull/13813

**Comment by davidlange6** [comment]
+1

**Comment by Dr15Jones** [comment]
It looks like this pull request is causing the crashes in the IB Release Validation Jobs: https://cms-sw.github.io/relvalLogDetail.html#slc6_amd64_gcc493;CMSSW_8_1_X_2016-05-11-2300

The traceback from the crash is 
```
#5  0x00007f93a367bf7d in hcaldqm::RawRunSummary::endJob(DQMStore::IBooker&, DQMStore::IGetter&) () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_1_X_2016-05-11-2300/lib/slc6_amd64_gcc493/libDQMHcalTasks.so
#6  0x00007f93a36f6aff in HcalOfflineHarvesting::_dqmEndJob(DQMStore::IBooker&, DQMStore::IGetter&) () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_1_X_2016-05-11-2300/lib/slc6_amd64_gcc493/pluginDQMHcalTasksAuto.so
#7  0x00007f93f97ac0d7 in DQMEDHarvester::endJob() () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_1_X_2016-05-11-2300/lib/slc6_amd64_gcc493/libDQMServicesCore.so
#8  0x00007f9400904df8 in edm::Worker::endJob() () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_1_X_2016-05-11-2300/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#9  0x00007f9400906790 in edm::WorkerManager::endJob(edm::ExceptionCollector&) () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_1_X_2016-05-11-2300/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#10 0x00007f94008bf755 in edm::Schedule::endJob(edm::ExceptionCollector&) () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_1_X_2016-05-11-2300/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#11 0x00007f9400836f91 in edm::EventProcessor::endJob() () from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_1_X_2016-05-11-2300/lib/slc6_amd64_gcc493/libFWCoreFramework.so
#12 0x000000000040ceea in main::{lambda()#1}::operator()() const ()
#13 0x000000000040b406 in main ()
```

**Comment by dmitrijus** [comment]
@vkhristenko can you please take care of this?

**Comment by vkhristenko** [comment]
@deguio , @dmitrijus , @Dr15Jones 

There are 2 separate issues in here. 

- Workflows 134.7** and 134.7** Already fixed. pushed. 1 statement should've been included...

- Workflow 4.77 
This workflow contains Harvesting over 2 runs - you can see that from logs of step3 for instance where the DQMEDAnalyzers are run for step3, but harvesting will get applied to that guy in a similar manner. Even though I misread https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkOneModuleInterface slightly.... it was guaranteed that HARVESTING step is done only over 1 run always for the following time period [Construction of Harvestor, Destruction of Harvestor].  I'm not sure how that is imposed... Because from the description of edm::one you can have (if you have WatchRuns) harvesting done over many runs...

**Since this was a guaranty, all of my Harvestors are designed to have harvesting being done only over 1 Run from Construction of the object until the destruction!** I've pushed an update which would be a temporary solution and it does pass 4.77 - basically just clear vectors per each beginRun transition...

my apology

VK

**Comment by dmitrijus** [comment]
I saw the fix for 8_0_X - but create another PR for 8_1_X, on the IB containing this. 
I will accept it - and hopefully make IBs valid again.

**Comment by deguio** [comment]
let me try to clarify:
- in prod at the moment there are no WFs that run the harvesting step on multiple runs (neither in relvals nor in prompt)
- in the future it would be nice to have this possibility supported (multiRun harvesting) so I would propose to extend the use case to cover multi run processing as well
- the skims are a special case due to the way in which they are processed in the RECO-DQM step:
   - each file could contain multiple runs
   - in prod at the harvesting N separate jobs are potentially run on the same file where N is the number of runs contained in that file
   - the LS from a given run are selected at the source where a filter is configured by WMAgent
   - even if only single runs are processed in each harvesting job, endRun - beginRun transitions could be triggered anyway during the processing.

In summary a design which assumes that the beginRun transition is called only once per harvesting job conclusion it is not safe.


I include few experts who might want to comment/correct me if I am wrong.
@cerminar @franzoni @fabozzi

**Comment by Dr15Jones** [comment]
From the framework perspective, modules should always be able to handle multiple Runs or LuminosityBlocks.
